druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java

```diff
@@ -604,30 +798,36 @@ private static boolean validSortLimit(Sort sort, DruidQuery query) {
         // offset not supported by Druid
         return false;
       }
-      if (query.getTopNode() instanceof Aggregate) {
-        final Aggregate topAgg = (Aggregate) query.getTopNode();
-        final ImmutableBitSet.Builder positionsReferenced = ImmutableBitSet.builder();
-        for (RelFieldCollation col : sort.collation.getFieldCollations()) {
-          int idx = col.getFieldIndex();
-          if (idx >= topAgg.getGroupCount()) {
-            continue;
-          }
-          // has the indexes of the columns used for sorts
-          positionsReferenced.set(topAgg.getGroupSet().nth(idx));
-        }
-        // Case it is a timeseries query
-        if (checkIsFlooringTimestampRefOnQuery(topAgg.getGroupSet(), topAgg.getInput(), query)
-            && topAgg.getGroupCount() == 1) {
-          // do not push if it has a limit or more than one sort key or we have sort by
-          // metric/dimension
-          return !RelOptUtil.isLimit(sort) && sort.collation.getFieldCollations().size() == 1
-              && checkTimestampRefOnQuery(positionsReferenced.build(), topAgg.getInput(), query);
+      // Use a different logic to push down Sort RelNode because the top node could be a Project now
+      RelNode topNode = query.getTopNode();
+      Aggregate topAgg;
+      if (topNode instanceof Project && ((Project) topNode).getInput() instanceof Aggregate) {
+        topAgg = (Aggregate) ((Project) topNode).getInput();
+      } else if (topNode instanceof Aggregate) {
+        topAgg = (Aggregate) topNode;
+      } else {
+        // If it is going to be a Druid select operator, we push the limit if
+        // it does not contain a sort specification (required by Druid)
+        return RelOptUtil.isPureLimit(sort);
+      }
+      final ImmutableBitSet.Builder positionsReferenced = ImmutableBitSet.builder();
+      for (RelFieldCollation col : sort.collation.getFieldCollations()) {
+        int idx = col.getFieldIndex();
+        if (idx >= topAgg.getGroupCount()) {
```

Comment:
I read the logic of `DruidSortRule`. It checks the type of Druid Query (GroupBy, Timeseries, TopN, etc.) by checking the input of aggregate in the druid query. The project after aggregate that is pushed in can only have input reference or math operation on numeric columns. Even if it changes the order of column, the type of druid query will not be changed and moreover we can still check the type of DruidQuery by checking aggregate in the Druid Query, so Sort will be still valid to be pushed in, since the sort get pushed in will take the last relNode in DQ as input. In the rendering time, the fieldNames of DQ already set correctly before the render of Sort Node, so it will get the right field names to render correct JsonCollation.

For other cases that post project does not get pushed in, the rule will be identical to the rule in before, so I think complexing checking is not necessary. Will add more test cases to verify it.